File: files/CH-061.py
import pandas as pd

input1 = pd.read_excel("CH-061 Sales per customer.xlsx", usecols="B:D", skiprows=1)
input2 = pd.read_excel("CH-061 Sales per customer.xlsx", usecols="F:G", skiprows=1, nrows = 6)
test = pd.read_excel("CH-061 Sales per customer.xlsx", usecols="I:J", skiprows=1, nrows = 8)\
    .sort_values(by="Sales", ascending=False)\
    .reset_index(drop=True)

def find_latest_id(id, changes):
    new_id = changes.loc[changes['OLD ID'] == id, 'New ID'].values
    if len(new_id) == 0:
        return id
    else:
        return find_latest_id(new_id[0], changes)

transactions = input1.copy()
transactions['Customer'] = transactions['Customer ID'].apply(lambda x: find_latest_id(x, input2))
transactions = transactions.groupby('Customer').agg({'Quantity': 'sum'})\
    .reset_index().rename(columns={'Quantity': 'Sales'})\
    .sort_values(by='Sales', ascending=False).reset_index(drop=True)

print(transactions.equals(test)) # True

File: files/000-099/061/CH-061.R
library(tidyverse)
library(readxl)

input1 = read_excel("files/CH-061 Sales per customer.xlsx", range = "B2:D36")
input2 = read_excel("files/CH-061 Sales per customer.xlsx", range = "F2:G8")
test   = read_excel("files/CH-061 Sales per customer.xlsx", range = "I2:J10") %>%
  arrange(desc(Sales))

find_latest_id <- function(id, changes) {
  new_id <- changes %>% filter(`OLD ID` == id) %>% pull(`New ID`)
  if (length(new_id) == 0) {
    return(id)
  } else {
    return(find_latest_id(new_id, changes))
  }
}

transactions <- input1 %>%
  mutate(Customer = map_chr(`Customer ID`, find_latest_id, input2)) %>%
  summarise(Sales = sum(Quantity), .by = Customer) %>%
  arrange(desc(Sales))

identical(test, transactions)
# [1] TRUE